packages/sw-background-sync-queue/src/lib/requestManager.js

```diff
@@ -0,0 +1,136 @@
+let globalCallbacks = {};
+
+/**
+ * Class to handle all the request related
+ * transformations, replaying, event handling
+ * broadcasting back to controlled pages etc.
+ * @class RequestManager
+ */
+class RequestManager {
+	/**
+	 * Initializes the request manager
+	 * stores the callbacks object, maintins global config and
+	 * attaches event handler
+	 * @param {Object} config config that gives values like
+	 * maxAge of requests in iDB
+	 * @param {Object} callbacks object that contains function
+	 * ref for various lifecycle callbacks
+	 *
+	 * @memberOf RequestManager
+	 */
+	async initialize(config, callbacks) {
+		this.attachSyncHandler();
+		globalConfig = Object.assign({}, initializationDefaults, config);
+		globalCallbacks = callbacks;
+		await queue.initialize( globalConfig );
+		await queue.cleanupQueue();
+	}
+
+	/**
+	 * attaches sync handler to replay requests when
+	 * sync event is fired
+	 *
+	 * @memberOf RequestManager
+	 */
+	attachSyncHandler() {
+		self.addEventListener('sync', (event) => {
+			event.waitUntil(this.replayRequests());
+		});
+	}
+
+	/**
+	 * function to start playing requests
+	 * in sequence
+	 * @return {Promise} promise that will resolve whenn all requests complete
+	 *
+	 * @memberOf RequestManager
+	 */
+	replayRequests() {
+		if (queue.getTotalTasks() < 1) {
+			return;
+		}
+		return this.doFetch(0);
+	}
+
+	/**
+	 * takes starting index and play all then
+	 * requests from that index onwards till the end
+	 *
+	 * @param {int} index starting point of playing
+	 * @return {void}
+	 *
+	 * @memberOf RequestManager
+	 */
+	async doFetch(index) {
+		let reqData = await queue.getRequestFromQueueAtIndex(index);
+		let hash = queue.getHash(index);
+		// exit point
+		if(!reqData) {
+			return;
+		}
+
+		// proceed if response do not already exist
+		if(reqData.response) {
+			return this.doFetch(index + 1);
+		}
+
+		let request = new Request(reqData.request.url, {
+			method: reqData.request.method,
+			headers: JSON.parse(reqData.request.headers),
+		});
+
+		if(request.method === 'POST') {
+			request.body = reqData.request.body;
+		}
+
+		let reqClone = request.clone();
+		return fetch( request )
+			.then( (response) => {
+				responseManager.putResponse(hash, reqData, response.clone());
+				globalCallbacks.onRetrySuccess
+					&& globalCallbacks.onRetrySuccess(hash, reqClone, response);
+				return this.doFetch(index + 1);
+			})
+			.catch((err) => {
+				globalCallbacks.onRetryFailure
+					&& globalCallbacks.onRetryFailure(hash, reqClone, err);
+				return this.doFetch( index + 1);
+			});
+	}
+
+	/**
+	 * takes a request and gives back JSON object that is storable in IDB
+	 *
+	 * @param {Request} request request object to transform
+	 * into iDB storable object
+	 * @param {Object} config config object to be
+	 * stored along inthe iDB
+	 * @return {Object} indexable object for iDB
+	 *
+	 * @memberOf RequestManager
+	 */
+	async getQueueableRequest(request, config) {
```

Comment:
This is similar to many of the other methods, so I'm not going to repeat the same thing for each one. But it's basically a static method that doesn't really have a connection to the overall `RequestManager` class. It doesn't make use of any class-specific state information.

For functions that will always be static helpers, I'd recommend just putting them inside their own modules not even attempting to wrap them in a class, e.g. https://github.com/GoogleChrome/sw-helpers/blob/030d824d583dbe4fb4bfd59a606caad667740287/packages/sw-broadcast-cache-update/src/lib/broadcast-update.js#L55